File: 2024/day01/main.py
import argparse


def read_input(filename):
    with open(filename, 'r') as input:
        left, right = zip(*(map(int, line.split()) for line in input))
    return list(left), list(right)


def part1(left, right):
    return sum(abs(r-l) for l, r in zip(sorted(left), sorted(right)))


def part2(left, right):
    right_ct = {r: right.count(r) for r in set(right)}
    return sum(l * right_ct.get(l, 0) for l in left)


if __name__ == '__main__':
    parser = argparse.ArgumentParser("main.py")
    parser.add_argument("filename", nargs='?',
                        help="input filename", default="input.txt")
    args = parser.parse_args()

    left, right = read_input(args.filename)

    sort_dist = part1(left, right)
    print(f"Part 1: sort_dist = {sort_dist}")

    similarity_score = part2(left, right)
    print(f"Part2: similarity_score = {similarity_score}")

File: 2024/day01/main.rs
use std::collections::HashMap;
use std::{env, fs};

fn read_input(filename: &str) -> (Vec<u32>, Vec<u32>) {
    fs::read_to_string(filename).expect("Unable to open file")
        .lines()
        .map(|line| {
            let parts: Vec<_> = line.split_whitespace().collect();
            let l = parts[0].parse::<u32>().unwrap();
            let r = parts[1].parse::<u32>().unwrap();
            (l, r)
        })
        .unzip()
}

fn part1(left: &[u32], right: &[u32]) -> u32 {
    left.iter().zip(right.iter())
        .map(|(l, r)| l.abs_diff(*r))
        .sum()
}

fn part2(left: &[u32], right: &[u32]) -> u32 {
    let mut right_ct = HashMap::new();

    for &r in right {
        *right_ct.entry(r).or_insert(0) += 1;
    }

    left.iter()
        .map(|&l| l * *right_ct.get(&l).unwrap_or(&0))
        .sum()
}

fn main() {
    let mut filename = "input.txt";
    let args: Vec<String> = env::args().collect();
    if args.len() == 2 {
        filename = &args[1];
    }

    let (mut left, mut right) = read_input(filename);
    left.sort();
    right.sort();

    println!("Part1: sort_dist = {}", part1(&left, &right));
    println!("Part2: similarity_score = {}", part2(&left, &right));
}
